### PR TITLE
adjusting regulating when modifying phase tap changer

### PR DESCRIPTION
**Pull request by EtienneLt**

### CODE REVIEW

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 1]
```diff
@@ -280,10 +280,13 @@ public static void processPhaseTapRegulation(PhaseTapChanger phaseTapChanger,
...
-        if (isModification) {
-            phaseTapChanger.setRegulating(true);
-        } else {
-            phaseTapChangerAdder.setRegulating(true);
+        ReportNode regulatingReport = ModificationUtils.getInstance().applyElementaryModificationsAndReturnReport(
+            isModification ? phaseTapChanger::setRegulating
+                : phaseTapChangerAdder::setRegulating,
+            isModification ? phaseTapChanger::isRegulating : () -> null,
+            phaseTapChangerInfos.getRegulating(), "Regulating", 2);
+        if (regulatingReport != null) {
```
Comment (Mathieu-Deharbe) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:288]:
```suggestion
        if (regulatingReport != null && regulationReports != null) {
```

**Review by Mathieu-Deharbe — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java [hunk 2]
```diff
@@ -280,10 +280,13 @@ public static void processPhaseTapRegulation(PhaseTapChanger phaseTapChanger,
...
 
-        if (isModification) {
-            phaseTapChanger.setRegulating(true);
-        } else {
-            phaseTapChangerAdder.setRegulating(true);
+        ReportNode regulatingReport = ModificationUtils.getInstance().applyElementaryModificationsAndReturnReport(
+            isModification ? phaseTapChanger::setRegulating
+                : phaseTapChangerAdder::setRegulating,
+            isModification ? phaseTapChanger::isRegulating : () -> null,
+            phaseTapChangerInfos.getRegulating(), "Regulating", 2);
```
Comment (Mathieu-Deharbe) [inline src/main/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModification.java:287]: You will probably have to pass Regulating as a parameter now.

**Review by Mathieu-Deharbe — CHANGES_REQUESTED** [review]
Doesn't compile anymore because of the recent merge of my refacto (that you reviewed) : https://github.com/gridsuite/network-modification-server/pull/524

Also can you please comment the ticket with your answers to the questions asked by the PO at the end of the orignal ticket ? I cannot understand if the issue is solved just with your code.

**Review by Mathieu-Deharbe — APPROVED** [review]
Small remark : I wonder how this "regulating" value is handled in this network modification now that we don't automatically set it to true. I didn't find how to modify it in the current interface.

Might be good to see this with a PO.

**Comment by thangqp** [comment]
> Small remark : I wonder how this "regulating" value is handled in this network modification now that we don't automatically set it to true. I didn't find how to modify it in the current interface.
> 
> Might be good to see this with a PO.

Share the same question????

**Review by Mathieu-Deharbe — CHANGES_REQUESTED** [review]
I wrote a lot about the regulation mode. If this is not clear just ask me.

I also got a weird bug but I don't see how it could be caused by your code (even if I only get it on your branch) so I don't write it for now. If you wan't to see just ask. I think I m going to look a bit more into it.

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java [hunk 3]
```diff
@@ -576,6 +577,215 @@ public void testPhaseTapChangerModification() throws Exception {
...
+            .phaseTapChanger(PhaseTapChangerModificationInfos.builder()
+                .enabled(new AttributeModification<>(true, OperationType.SET))
+                .regulationMode(new AttributeModification<>(PhaseTapChanger.RegulationMode.CURRENT_LIMITER, OperationType.SET))
+                .regulationValue(new AttributeModification<>(10.0, OperationType.SET))
+                .build())
+            .build();
+
+        String modificationToModifyJson = mapper.writeValueAsString(phaseTapChangerCreation);
+        mockMvc.perform(post(getNetworkModificationUri()).content(modificationToModifyJson).contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk()).andReturn();
```
Comment (thangqp) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java:637]:
```suggestion
            .andExpect(status().isOk());
```
result is not used, do not need .andReturn();

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java [hunk 4]
```diff
@@ -576,6 +577,215 @@ public void testPhaseTapChangerModification() throws Exception {
...
+        assertEquals(0.0, phaseTapChanger.getTargetDeadband());
+        assertEquals(10.0, phaseTapChanger.getRegulationValue());
+
+        // modification 2  : CURRENT_LIMITER -> FIXED_TAP
+        phaseTapChangerCreation.getPhaseTapChanger().setRegulationMode(new AttributeModification<>(PhaseTapChanger.RegulationMode.FIXED_TAP, OperationType.SET));
+        phaseTapChangerCreation.getPhaseTapChanger().setRegulationValue(null);
+
+        String modificationToModifyJson2 = mapper.writeValueAsString(phaseTapChangerCreation);
+        mockMvc.perform(post(getNetworkModificationUri()).content(modificationToModifyJson2).contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk()).andReturn();
```
Comment (thangqp) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java:653]:
```suggestion
            .andExpect(status().isOk());
```

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java [hunk 5]
```diff
@@ -576,6 +577,215 @@ public void testPhaseTapChangerModification() throws Exception {
...
+        assertEquals(10.0, phaseTapChanger.getRegulationValue());
+        assertFalse(phaseTapChanger.isRegulating());
+
+        // modification 3   : FIXED_TAP -> ACTIVE_POWER_CONTROL
+        phaseTapChangerCreation.getPhaseTapChanger().setRegulationMode(new AttributeModification<>(PhaseTapChanger.RegulationMode.ACTIVE_POWER_CONTROL, OperationType.SET));
+        phaseTapChangerCreation.getPhaseTapChanger().setTargetDeadband(new AttributeModification<>(1.0, OperationType.SET));
+
+        String modificationToModifyJson3 = mapper.writeValueAsString(phaseTapChangerCreation);
+        mockMvc.perform(post(getNetworkModificationUri()).content(modificationToModifyJson3).contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk()).andReturn();
```
Comment (thangqp) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java:668]:
```suggestion
            .andExpect(status().isOk());
```

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java [hunk 6]
```diff
@@ -576,6 +577,215 @@ public void testPhaseTapChangerModification() throws Exception {
...
+        assertTrue(phaseTapChanger.isRegulating());
+
+        // modification 4 : ACTIVE_POWER_CONTROL -> CURRENT_LIMITER
+        phaseTapChangerCreation.getPhaseTapChanger().setRegulationMode(new AttributeModification<>(PhaseTapChanger.RegulationMode.CURRENT_LIMITER, OperationType.SET));
+        phaseTapChangerCreation.getPhaseTapChanger().setRegulationValue(new AttributeModification<>(8.0, OperationType.SET));
+        phaseTapChangerCreation.getPhaseTapChanger().setTargetDeadband(new AttributeModification<>(2.0, OperationType.SET));
+
+        String modificationToModifyJson4 = mapper.writeValueAsString(phaseTapChangerCreation);
+        mockMvc.perform(post(getNetworkModificationUri()).content(modificationToModifyJson4).contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk()).andReturn();
```
Comment (thangqp) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java:685]:
```suggestion
            .andExpect(status().isOk());
```

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java [hunk 7]
```diff
@@ -576,6 +577,215 @@ public void testPhaseTapChangerModification() throws Exception {
...
+        assertTrue(phaseTapChanger.isRegulating());
+
+        // modification 5 : CURRENT_LIMITER -> FIX_TAP
+        phaseTapChangerCreation.getPhaseTapChanger().setRegulationMode(new AttributeModification<>(PhaseTapChanger.RegulationMode.FIXED_TAP, OperationType.SET));
+        phaseTapChangerCreation.getPhaseTapChanger().setRegulationValue(null);
+        phaseTapChangerCreation.getPhaseTapChanger().setTargetDeadband(null);
+
+        String modificationToModifyJson5 = mapper.writeValueAsString(phaseTapChangerCreation);
+        mockMvc.perform(post(getNetworkModificationUri()).content(modificationToModifyJson5).contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk()).andReturn();
```
Comment (thangqp) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java:702]:
```suggestion
            .andExpect(status().isOk());
```

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java [hunk 8]
```diff
@@ -576,6 +577,215 @@ public void testPhaseTapChangerModification() throws Exception {
...
+                .regulationMode(new AttributeModification<>(PhaseTapChanger.RegulationMode.ACTIVE_POWER_CONTROL, OperationType.SET))
+                .build())
+            .build();
+
+        String modificationToModifyJson1 = mapper.writeValueAsString(phaseTapChangerCreation);
+
+        // modification 1 assert
+        MvcResult result = mockMvc.perform(post(getNetworkModificationUri())
+                .content(modificationToModifyJson1).contentType(MediaType.APPLICATION_JSON))
+            .andExpect(status().isOk()).andReturn();
```
Comment (thangqp) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java:735]:
```suggestion
        mockMvc.perform(post(getNetworkModificationUri())
                .content(modificationToModifyJson1).contentType(MediaType.APPLICATION_JSON))
            .andExpect(status().isOk());
```

File: src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java [hunk 9]
```diff
@@ -576,6 +577,215 @@ public void testPhaseTapChangerModification() throws Exception {
...
+        assertEquals("{\"applicationStatus\":\"WITH_ERRORS\",\"lastGroupApplicationStatus\":\"WITH_ERRORS\",\"networkImpacts\":[],\"impactedSubstationsIds\":[]}",
+            result.getResponse().getContentAsString());
```
Comment (thangqp) [inline src/test/java/org/gridsuite/modification/server/modifications/TwoWindingsTransformerModificationTest.java:737]:
```suggestion
            assertLogMessage(new NetworkModificationException(MODIFY_TWO_WINDINGS_TRANSFORMER_ERROR, "Regulation value is missing, phase tap changer can not regulate").getMessage(),
            phaseTapChangerCreation.getErrorType().name(), reportService);
```

In my opinion, an assert on result is less interesting than an assert on log

**Review by thangqp — COMMENTED** [review]
Could you add an additional test method for the creation of extension, pls? with name e.g. testPhaseTapChangerRegulationCreation(), I see in the code coverage, it does not yet cover some required modification

**Review by thangqp — APPROVED** [review]
Test OK, Code OK

**Review by Mathieu-Deharbe — APPROVED** [review]

**Pull request merged**